### PR TITLE
Load plugins for apis server.

### PR DESCRIPTION
**Pull request by absoludity**
### Description of the change

Follows #2789 adding two plugins with unimplemented servers for the `packages.GetAvailablePackages` call together with the functionality from the POC to load the plugins.

The main difference from the POC is the file layout, with the proto and generated files separated from the actual plugin implementation (I'll update the README with this info).

Next up:
* Implement the `GetAvailablePlugins` call now that plugins are loaded
* Add Dockerfile and update the dev chart to deploy the service behind a feature flag
* Implement and test `GetAvailablePackages` for kapp-controller.

### Applicable issues

  - ref #2773

### CODE REVIEW

File: cmd/kubeapps-apis/README.md [hunk 1]
```diff
@@ -16,6 +16,17 @@ Together, this enables the best of both worlds: a well known and used Interface
...
 Finally, we've also chosen to use the [buf](https://buf.build/) tool for generating the code from the proto files. In the past we've used `protoc` (proto buffer compiler) and its extensions directly, but `buf` allows you to specify a simple yaml config instead, and also provides a `lint` command to ensure that your choice of API structure follows best practise, as well as ensuring you're aware when you break backwards compatability.
 
+## Plug-able
+
+The kubeapps-apis service uses the standard [go plugin package](https://golang.org/pkg/plugin/) to be able to load api plugins at run-time.
+
+Each plugin consists of 2 source files (and some generated files):
+
+* A `.proto` file defining the service that uses the messages defined in relevant part of kubeappsapis.core,
+* A `main.go` that compiles to an .so file for that plugin. This `main.go` has two public functions: one to register the plugin with a GRPC server and one to register the plugin for the http handler as well as the implementation for the server. This may be split into further modules as the complexity of the plugin grows.
```
Comment (antgamdia) [inline cmd/kubeapps-apis/README.md:26]: Perhaps obvious, but I'd say the `.go` files are is under `./plugins/<pluginName>` whereas the `.proto` are under `./proto/kubeappsapis/plugins/<pluginName>`

File: cmd/kubeapps-apis/server/server.go [hunk 2]
```diff
@@ -88,3 +109,88 @@ type gwHandlerArgs struct {
...
+		}
+
+		log.Infof("Successfully registered plugin %q", pluginPath)
+	}
+	return nil
+}
+
+// registerGRPC finds and calls the required function for registering the plugin for the GRPC server.
+func registerGRPC(p *plugin.Plugin, pluginPath string, registrar grpc.ServiceRegistrar) error {
+	grpcRegFn, err := p.Lookup(grpcRegisterFunction)
```
Comment (antgamdia) [inline cmd/kubeapps-apis/server/server.go:136]: Interesting usage of reflection here to be able to register the plugging in both grpc and HTTP server.

File: cmd/kubeapps-apis/server/server.go [hunk 3]
```diff
@@ -88,3 +109,88 @@ type gwHandlerArgs struct {
...
+		var dummyFn gatewayRegisterFunctionType = func(context.Context, *runtime.ServeMux, string, []grpc.DialOption) error { return nil }
+		return fmt.Errorf("unable to use %q in plugin %q due to mismatched signature.\nwant: %T\ngot: %T", gatewayRegisterFunction, pluginPath, dummyFn, gwRegFn)
+	}
+	return gwfn(gwArgs.ctx, gwArgs.mux, gwArgs.addr, gwArgs.dialOptions)
+}
+
+// listSOFiles returns the absolute paths of all .so files found in any of the provided plugin directories.
+//
+// pluginDirs can be relative to the current directory or absolute.
+func listSOFiles(fsys fs.FS, pluginDirs []string) ([]string, error) {
```
Comment (antgamdia) [inline cmd/kubeapps-apis/server/server.go:170]: I guess during these early dev stages no, but do you foresee any security concern loading arbitrary so files? Should we allowlisting the plugins in the values.yaml instead?

File: cmd/kubeapps-apis/server/server_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+			if got, want := got, tc.pluginFilenames; !cmp.Equal(want, got) {
+				t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got))
+			}
+
+		})
+	}
+}
+
+func createTestFS(t *testing.T, filenames []string) fstest.MapFS {
```
Comment (antgamdia) [inline cmd/kubeapps-apis/server/server_test.go:79]: Is this the in-memory file system, that new test feature in Go 1.16, isn't it?

**Review by antgamdia — APPROVED** [review]
Interesting dynamic load of the plugins. Perhaps at some point we may think of declaring the plugins in the values-yaml instead of relying on an arbitrary list of so files.
Thank you for this clear PR, very understandable and targeted!

File: cmd/kubeapps-apis/README.md (hunk 1 above)
Comment (absoludity) [reply cmd/kubeapps-apis/README.md]: Done, thanks.

File: cmd/kubeapps-apis/server/server.go (hunk 2 above)
Comment (absoludity) [reply cmd/kubeapps-apis/server/server.go:136]: This is how the `plugin` package recommends calling the functions. The only non-standard bit is the `dummyFn` to try to get a useful error message.

File: cmd/kubeapps-apis/server/server.go (hunk 3 above)
Comment (absoludity) [reply cmd/kubeapps-apis/server/server.go:170]: I don't think so, in that they won't load and we'll error, unless they have the required functions for the plugin, in which case, it is a plugin in the directory specified by the caller, whether it's evil or not. Not sure, need to think more.

File: cmd/kubeapps-apis/server/server_test.go (hunk 4 above)
Comment (absoludity) [reply cmd/kubeapps-apis/server/server_test.go:79]: Yeah, I was looking to test this and found it. In the past I've had to create actual files in a /tmp dir for tests, so this was nice to try out :)

**Review by absoludity — COMMENTED** [review]
Thanks Antonio!

**Pull request merged**